### PR TITLE
Vmunier/cors

### PR DESCRIPTION
**Pull request by vmunier**
* Add a Play filter which adds CORS headers to every HTTP request.
* Add a Cors controller which can be used with `OPTIONS  /*all`

This commit illustrates what I wanted to do initially https://github.com/vmunier/playpen/commit/4ee184fa320b327311b460fb0b56d21956d74561, i.e. being able to describe the allowed methods for each route. I am happy to remove this commit if needed.

### CODE REVIEW

File: src/main/scala/playpen/cors/filters/CorsFilter.scala [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package playpen.cors.filters
+
+import play.api.mvc.{EssentialFilter, EssentialAction, RequestHeader}
+import playpen.cors.CorsHeaders
+import scala.concurrent.ExecutionContext.Implicits.global
```
Comment (dwijnand) [inline src/main/scala/playpen/cors/filters/CorsFilter.scala]: I think we want to use `import play.api.libs.concurrent.Execution.Implicits.defaultContext` here, in this Play walled garden

File: src/main/scala/playpen/cors/CorsHeaders.scala [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package playpen.cors
+import play.api.Play._
+import play.api.mvc.RequestHeader
+
+object CorsHeaders {
+  lazy val allowedOrigins = current.configuration.getStringSeq("cors.allowed.origins").getOrElse(Seq())
+  lazy val allowedMethods = current.configuration.getStringSeq("cors.allowed.methods").getOrElse(Seq("POST", "GET", "OPTIONS", "PUT", "DELETE"))
```
Comment (dwijnand) [inline src/main/scala/playpen/cors/CorsHeaders.scala]: I think we should scope these for playpen, so `playpen.cors.allowed.origins` etc, `cors` is too general for a library IMO

File: src/main/scala/playpen/cors/CorsHeaders.scala [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+  def headers(request: RequestHeader): Seq[(String, String)] = request match {
+    case CorsParams(params) if allowedOrigins.isEmpty | allowedOrigins.exists(params.origin.endsWith(_)) =>
```
Comment (dwijnand) [inline src/main/scala/playpen/cors/CorsHeaders.scala]: I think we should be secure-by-default: if you've configured the CorsFilter nothing will be "Access-Control-Allow-Origin" unless you configure `playpen.cors.allowed.origins`

**Comment by dwijnand** [comment]
So:
* Looks like a great addition
* I'm not too savvy on CORS, so another pair of eyes would be great
* Added some comments.

File: src/main/scala/playpen/cors/CorsHeaders.scala (hunk 3 above)
Comment (vmunier) [inline src/main/scala/playpen/cors/CorsHeaders.scala]: Ok, I think we should go for `if allowedOrigins.exists(orig => orig == "*" | params.origin.endsWith(orig))`

Comment (dwijnand) [inline src/main/scala/playpen/cors/CorsHeaders.scala]: If that makes sense for CORS then sure, but then if you've configured `playpen.cors.allowed.origins` to `*` would you want to respond maybe with `Access-Control-Allow-Origin: *`?

Comment (vmunier) [inline src/main/scala/playpen/cors/CorsHeaders.scala]: The behaviour would be the same, you've requested `example.com`, you always get `example.com` back. But it's maybe clearer to return `*`, I am taking suggestions on how to change the code to do that :)

**Comment by vmunier** [comment]
@dwijnand I addressed your comments. Do you want to me to squash the commits before the PR gets merged?

File: src/main/scala/playpen/cors/CorsHeaders.scala (hunk 3 above)
Comment (dwijnand) [inline src/main/scala/playpen/cors/CorsHeaders.scala]: I guess that's also a better safety-by-default (instead of `*`).

**Comment by dwijnand** [comment]
No, that's ok. Thanks for the feature man.

**Pull request merged**